### PR TITLE
BF: default location for non-listed endpoint

### LINKED ISSUE #669: Location Constraint Info is invalid.
Anyone seeing this error when trying to upload an object?  Getting it with CloudBerry and S3Browser and latest build (7 days ago.)

`{"name":"S3","clientIP":"::ffff:192.168.20.1","clientPort":19601,"httpMethod":"PUT","httpURL":"/kscott1/A79F758FD4.DIF","time":1490722748203,"req_id":"7d4208edd8b87f01cd8c","level":"info","message":"received request","hostname":"22370bf410a8","pid":49}
{"name":"S3","bucketName":"kscott1","objectKey":"A79F758FD4.DIF","bytesReceived":2052592,"bodyLength":2052592,"accessKey":"accessKey1","bytesSent":200,"clientIP":"::ffff:192.168.20.1","clientPort":19601,"httpMethod":"PUT","httpURL":"/kscott1/A79F758FD4.DIF","contentLength":2052592,"httpCode":400,"time":1490722748206,"req_id":"7d4208edd8b87f01cd8c","elapsed_ms":3.303361,"level":"info","message":"responded with error XML","hostname":"22370bf410a8","pid":49}
`

### PR DESCRIPTION
**Pull request by nicolas2bert**
FIXES #669

### CODE REVIEW

File: lib/api/apiUtils/object/BackendInfo.js [hunk 1]
```diff
@@ -131,6 +141,7 @@ export class BackendInfo {
     * (1) objectLocationConstraint,
     * (2) bucketLocationConstraint,
     * (3) default locationConstraint for requestEndpoint
```
Comment (LaurenSpiegel) [inline lib/api/apiUtils/object/BackendInfo.js]: add to (3) if requestEndpoint is listed in restEndpoints in config.json

File: lib/api/apiUtils/object/BackendInfo.js [hunk 2]
```diff
@@ -50,6 +52,14 @@ export class BackendInfo {
         if (Object.keys(config.restEndpoints).indexOf(requestEndpoint) < 0) {
             log.trace('requestEndpoint does not match config restEndpoints',
                 { requestEndpoint });
+            if (config.backends.data === 'mem' ||
```
Comment (LaurenSpiegel) [inline lib/api/apiUtils/object/BackendInfo.js]: as discussed, please separate logic so we have (1) a function checking whether the requestEndpoint is in the restEndpoints, (2) a function checking whether config.backends.data is mem or file and (3) a function checking whether the value of the requestEndpoint in the restEndpoints is valid.

Comment (nicolas2bert) [inline lib/api/apiUtils/object/BackendInfo.js]: Done

**Comment by LaurenSpiegel** [comment]
Don't you have to update the isValidControllingBackendParam function?

**Comment by nicolas2bert** [comment]
Do you mean `controllingBackendParam` ? I updated `isValidRequestEndpoint` but it still has the same name so it does not change anything for `controllingBackendParam`.

File: lib/api/apiUtils/object/BackendInfo.js [hunk 3]
```diff
@@ -38,20 +38,28 @@ export class BackendInfo {
     }
 
     /**
-     * validate requestEndpoint against config
-     * if there is a mismatch between the request endpoint and what is in
-     * the config, this could cause a problem for setting the backend location
-     * for storing data
+     * validate that request endpoint is listed in the requestEndpoint config
```
Comment (LaurenSpiegel) [inline lib/api/apiUtils/object/BackendInfo.js]: restEndpoint

Comment (nicolas2bert) [inline lib/api/apiUtils/object/BackendInfo.js]: Done - thanks :)

File: lib/api/apiUtils/object/BackendInfo.js [hunk 4]
```diff
@@ -63,6 +71,42 @@ export class BackendInfo {
...
+                method: 'isMemOrFileBackend',
+            });
+            return true;
+        }
+        return false;
+    }
+
+    /**
+     * validate requestEndpoint against config or mem/file data backend
+     * - if there is a mismatch between the request endpoint and what is in
```
Comment (dora-korpar) [inline lib/api/apiUtils/object/BackendInfo.js]: I would change this to something like "if there is no match for the request endpoint in the config restEndpoints....."
just to make it a bit clearer

**Comment by nicolas2bert** [comment]
@ironman-machine ok

**Pull request merged**